### ISSUE TITLE
Video does not start with geckodriver

### ISSUE THREAD

**Issue by sky-joaonabais** [mozilla/geckodriver#755]
I'm working with automated tests and I'm trying to set up selenium grid for running tests in multiple environments.

I've successfully ran tests with chrome driver in Mac OS X 10.11 and Windows10.

Right now when I try to run a test feature, in a page with a video element, and the following message prompts on firefox (geckodriver):

> Firefox is installing components needed to play audio or video, please try again later and the video does not start.

I'm running selenium standalone both on the Hub(Mac OS X) and Node(Windows 10) of the selenium grid.

I've tried to create a profile and import it on the node config (windowsNodeConfig.json):

```
{
"capabilities": [
    (...)
    {
        "browserName": "firefox",
        "maxInstances": 5,
        "platform": "WINDOWS",
        "marionette": true,
        "acceptInsecureCerts": true,
        "webdriver.gecko.driver": "geckodriver.exe",
        "firefox_profile": "firefoxProfile"
    }
    (...)
],
"nodeTimeout": 180000,
"browserTimeout": 180,
"timeout": 180,
"newSessionWaitTimeout": 180,
"cleanUpCycle": 5000,
"firefoxProfileTemplate": "firefoxProfile"
}
```
Versions: Selenium standalone : 3.4.0
Geckodriver : 0.16.1
Firefox (Windows) : 53.0.3

I'm running the node as follows:

`$ java -jar selenium-server-standalone-3.4.0.jar -role node -nodeConfig windowsNodeConfig.json -hub http://<IP_ADDRESS>:4444`

Running the Hub as follows:

`$ java -jar selenium-hub/selenium-server-standalone-3.4.0.jar -role hub`

The firefox profile I've created has the Widevine Content Decryption Module addon installed. However on the node logs I can see the following line:

`JavaScript error: resource://gre/modules/AddonManager.jsm, line 2570: NS_ERROR_NOT_INITIALIZED: AddonManager is not initialized`

The same test feature runs in chrome.

I've created the firefox profile as described in: https://support.mozilla.org/en-US/kb/profile-manager-create-and-remove-firefox-profiles#w_creating-a-profile

Am I missing any config?

**Comment by andreastt** [comment]
> Right now when I try to run a test feature, in a page with a video element, and the following message prompts on firefox (geckodriver):
> 
>> Firefox is installing components needed to play audio or video, please try again later and the video does not start.

I think this indicates that the video is using DRM and that it needs to download the proprietary EME binary blob. Usually you have to allow Firefox to do this in about:preferences, but maybe you have that switched on.

We do explicitly disable plugins such as Flash from running with geckodriver due to issues with process crashes on Windows: https://github.com/mozilla/geckodriver/blob/master/src/prefs.rs#L195

Maybe you could try set `plugin.state.flash` to true.

> The firefox profile I've created has the Widevine Content Decryption Module addon installed. However on the node logs I can see the following line:
> 
> `JavaScript error: resource://gre/modules/AddonManager.jsm, line 2570: NS_ERROR_NOT_INITIALIZED: AddonManager is not initialized`

This is nothing to worry about as it does not originate from Marionette.

**Comment by sky-joaonabais** [comment]
> Firefox is installing components needed to play audio or video, please try again later and the video does not start.

is displayed because I'm trying to access a page with a <video> element using DRM, more precisely Widevine and I need the **Widevine Content Decryption Module** addon.

I am not trying to access a page with flash.

Example of video with widevine:
http://amp.azure.net/libs/amp/latest/samples/videotag_multiDRM_PlayReadyWidevineFairPlay_notoken.html

**Comment by AutomatedTester** [comment]
Having spoken to the media team about this there is nothing that we can do unfortunately. The best course of action is to create a [Firefox profile](https://support.mozilla.org/en-US/kb/profile-manager-create-and-remove-firefox-profiles). Navigate to the site, or any other site that would initiate the DRM download and wait for it to finish. 

Selenium allows you to then pass through profiles which are then given to GeckoDriver and we recreate that profile on disk and it should work from then on.

There is unfortunately nothing the GeckoDriver team can do to fix this as Firefox is using a 3rd party here to do the work and we can't control that.

**Comment by sky-joaonabais** [comment]
@AutomatedTester I did just that, can you take a look to the original message and tell me what I might be doing wrong?

Thanks.

**Comment by AutomatedTester** [comment]
@sky-joaonabais where is your code for passing in the profile? I only see your node configuration which won't make any difference for firefox profiles.

Please feel free to send questions to the [selenium user group](https://groups.google.com/forum/#!forum/selenium-users)

**Comment by sky-joaonabais** [comment]
@AutomatedTester I'm generating the profile as specified [here](https://support.mozilla.org/en-US/kb/profile-manager-create-and-remove-firefox-profiles#w_starting-the-profile-manager) and creating a base64 zipped version of the profile version:

`node_modules/.bin/firefox-profile -c firefoxProfile -b` (using [firefox-profile-jo](https://github.com/saadtazi/firefox-profile-js))

and 

`gzip profileBase64`

I'm importing it in `capabilities` object (using [webdriver.io](http://webdriver.io/guide/getstarted/configuration.html#desiredCapabilities):

```
[
(...)
{
  maxInstances: 5,
  browserName: 'firefox',
  marionette: true,
  acceptInsecureCerts: true,
  log: {
    level: 'trace'
  },
  'moz:firefoxOptions': {
    profile: 'profileBase64.gz'
  }
}
(...)
]
```

I get the following log (on selenium node):
```
INFO - Creating a new session for Capabilities [{rotatable=true, marionette=true, log={level=trace}, acceptInsecureCerts=true, loggingPrefs=org.openqa.selenium.logging.LoggingPreferences@6448c9d7, locationContextEnabled=true, browserName=firefox, javascriptEnabled=true, handlesAlerts=true, moz:firefoxOptions={profile=profileBase64.gz}, maxInstances=5, requestOrigins={name=webdriverio, version=4.6.2, url=http://webdriver.io}}]
```

So I'm assuming the profile is imported successfully. However I'm still seeing: 

> Firefox is installing components needed to play audio or video, please try again later and the video does not start.

Am I missing any config on the import of the profile?

**Comment by whimboo** [comment]
Due to a regression in geckodriver the addon manager is busted. See issue #702. Without it I doubt that plugins can successfully be downloaded. That's why you are seeing this failure.

A fix for the underlying issue has already landed for geckodriver, so I'm aiming a new release of it really soon. It should fix your problem.

**Comment by sky-joaonabais** [comment]
Thanks @whimboo I'll wait for the release.

**Comment by sky-joaonabais** [comment]
@whimboo any news on this? Thanks

**Comment by sky-joaonabais** [comment]
I've tested with
Firefox version: 54.0.1 (64-bit)
Ghecko driver: 0.16.1-x64

It is working seamlessly

**Comment by whimboo** [comment]
This is working with geckodriver 0.17 and upwards.

**Comment by sky-joaonabais** [comment]
@whimboo, This is working on Mac OS X but in Windows 10 I got the following:

```
15:36:40.582 INFO - Executing: [new session: Capabilities [{rotatable=true, marionette=true, log={level=trace}, loggingPrefs=org.openqa.selenium.logging.LoggingPreferences@844ea7b, webdriver.gecko.driver=drivers/geckodriver/Windows/v0.18.0/geckodriver.exe, handlesAlerts=true, moz:firefoxOptions={profile=profileWindowsBase64.zip}, platform=WINDOWS, requestOrigins={name=webdriverio, version=4.6.2, url=http://webdriver.io}, acceptInsecureCerts=true, locationContextEnabled=true, browserName=firefox, javascriptEnabled=true, maxInstances=5}]])
15:36:40.582 INFO - Creating a new session for Capabilities [{rotatable=true, marionette=true, log={level=trace}, loggingPrefs=org.openqa.selenium.logging.LoggingPreferences@844ea7b, webdriver.gecko.driver=drivers/geckodriver/Windows/v0.18.0/geckodriver.exe, handlesAlerts=true, moz:firefoxOptions={profile=profileWindowsBase64.zip}, platform=WINDOWS, requestOrigins={name=webdriverio, version=4.6.2, url=http://webdriver.io}, acceptInsecureCerts=true, locationContextEnabled=true, browserName=firefox, javascriptEnabled=true, maxInstances=5}]
1500907000738    geckodriver    INFO    Listening on 127.0.0.1:48333
1500907001234    geckodriver::marionette    INFO    Starting browser \\?\C:\Program Files (x86)\Mozilla Firefox\firefox.exe with args ["-marionette"]
1500907001576    addons.manager    DEBUG    Application has been upgraded
1500907001577    addons.manager    ERROR    startup failed: [Exception... "Component returned failure code: 0x80070057 (NS_ERROR_ILLEGAL_VALUE) [nsIFile.create]"  nsresult: "0x80070057 (NS_ERROR_ILLEGAL_VALUE)"  location: "JS frame :: resource://gre/modules/FileUtils.jsm :: FileUtils_getDir :: line 70"  data: no] Stack trace: FileUtils_getDir()@resource://gre/modules/FileUtils.jsm:70 < FileUtils_getFile()@resource://gre/modules/FileUtils.jsm:42 < validateBlocklist()@resource://gre/modules/AddonManager.jsm:674 < startup()@resource://gre/modules/AddonManager.jsm:837 < startup()@resource://gre/modules/AddonManager.jsm:3145 < observe()@resource://gre/components/addonManager.js:65
1500907001578    addons.manager    DEBUG    Completed startup sequence
JavaScript error: resource://gre/modules/AddonManager.jsm, line 1657: NS_ERROR_NOT_INITIALIZED: AddonManager is not initialized
[GFX1]: Potential driver version mismatch ignored due to missing DLLs igd10umd32 v= and igd10iumd32.dll v=0.0.0.0
*** Blocklist::_preloadBlocklistFile: blocklist is disabled
1500907003397    Marionette    INFO    Listening on port 52016
JavaScript error: resource://gre/modules/AddonManager.jsm, line 2572: NS_ERROR_NOT_INITIALIZED: AddonManager is not initialized
1500907004370    Marionette    WARN    TLS certificate errors will be ignored for this session
15:36:44.400 INFO - Detected dialect: W3C
```

Using geckodriver 0.18.0:

```
{
    maxInstances: 5,
    browserName: 'firefox',
    marionette: true,
    acceptInsecureCerts: true,
    log: {
        level: 'trace',
    },
    platform: 'WINDOWS',
    'webdriver.gecko.driver': 'drivers/geckodriver/Windows/v0.18.0/geckodriver.exe',
    // TODO: reate profile for MAC OS X
    'moz:firefoxOptions': {
        profile: 'profileWindowsBase64.zip',
    },
}
```

I've created the firefox profile using profile manager I've created the base64 version whit:
```
firefox-profile -c ek20n4kc.firefoxProfile --base64 profileWindowsBase64
```

And then I've zipped it.

Not sure If I'm missing anything here, but I'm still getting a `AddonManager is not initialized`.

I'm using Firefox 54.0.1

**Comment by whimboo** [comment]
I don't think that you are using geckodriver 0.18 because of the following path:

> \\?\C:\Program Files (x86)\Mozilla Firefox\firefox.exe

I fixed that in version 0.17 of geckodriver via issue #702.

**Comment by duncanleo** [comment]
Apologies for re-opening an old thread, but for anyone still needing to fix this problem, enabling the `media.eme.enabled` preference worked for me.

Sample Ruby code:
```ruby
options = Selenium::WebDriver::Firefox::Options.new
options.headless!
options.add_preference('media.eme.enabled', true)

driver = Selenium::WebDriver.for :firefox, options: options
```
